crates/bin/pd/src/migrate/mainnet3.rs

```diff
@@ -0,0 +1,103 @@
+//! Migration for shipping consensus-breaking IBC changes, updating
+//! IBC-related dependencies for light-client verification paths, via
+//! `tendermint-rs` and associated `ibc-types` deps.
+use cnidarium::{StateDelta, Storage};
+use jmt::RootHash;
+use penumbra_sdk_app::app::StateReadExt as _;
+use penumbra_sdk_app::app_version::migrate_app_version;
+use penumbra_sdk_governance::StateWriteExt;
+use penumbra_sdk_sct::component::clock::EpochManager;
+use penumbra_sdk_sct::component::clock::EpochRead;
+use std::path::PathBuf;
+use tracing::instrument;
+
+use crate::network::generate::NetworkConfig;
+
+/// Run the full migration, emitting a new genesis event, representing historical state.
+///
+/// This will have the effect of reinserting packets which had acknowledgements containing
+/// errors, and erroneously removed from state, as if the acknowledgements had contained successes.
+#[instrument]
+pub async fn migrate(
+    storage: Storage,
+    pd_home: PathBuf,
+    genesis_start: Option<tendermint::time::Time>,
+) -> anyhow::Result<()> {
+    // Setup:
+    let initial_state = storage.latest_snapshot();
+    let chain_id = initial_state.get_chain_id().await?;
+    let root_hash = initial_state
+        .root_hash()
+        .await
+        .expect("chain state has a root hash");
+    // We obtain the pre-upgrade hash solely to log it as a result.
+    let pre_upgrade_root_hash: RootHash = root_hash.into();
+    let pre_upgrade_height = initial_state
+        .get_block_height()
+        .await
+        .expect("chain state has a block height");
+    let post_upgrade_height = pre_upgrade_height.wrapping_add(1);
+
+    let mut delta = StateDelta::new(initial_state);
+    let (migration_duration, post_upgrade_root_hash) = {
+        let start_time = std::time::SystemTime::now();
+
+        migrate_app_version(&mut delta, 9).await?;
```

Comment:
should this be 9 or 10, i think 10 right?

Reply:
You're right, it should be 10. This is the part I always forget in writing no-op migrations. I've pushed another commit that changes L45 in `mainnet3.rs` to use 10, and also bumps the default app version for pd to be 10, too.